autoprotocol/instruction.py

```diff
@@ -24,15 +23,18 @@ def __init__(self, op, data):
         self.data = self._remove_empty_fields(data)
         self.__dict__.update(self.data)
 
-    def json(self):
-        """Return instruction object properly encoded as JSON for Autoprotocol.
+    def __repr__(self):
+        return "Instruction({}, {})".format(self.op, self.data)
 
+    def _as_AST(self):
+        """generates a JSON serializable representation of the Instruction
         Returns
         -------
-        str
-            Instruction object encoded as json string
+        dict
+            A JSON serializable representation of the Instruction
+
         """
-        return json.dumps(dict(op=self.op, **self.data), indent=2)
+        return dict(op=self.op, **self.data)
```

Comment:
hmm... could we still leave in the json method for backwards compatibility? We could wrap it around the `_as_AST` method if we want

Reply:
going to clean up docs, but as noted in conversation the `Instruction.json` method doesn't actually work when there are non JSON-serializeable objects referenced in its data. the most prevalent case of this is when an Instruction references a Container or other Autoprotocol Python object.

in general it would be nice to move towards more robust (de)serialization in the future.